rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/locations/impl/TreeSearch.java

```diff
@@ -0,0 +1,124 @@
+ */
+public class TreeSearch {
+    
+    /**
+     * Helper to test if a line/column position is within the scope of a source location's 
+     * begin line/column and end line/column scope. The first and the last lines are
+     * where the column information is used.
+     * 
+     * @param loc     location with line/column information that acts as a range
+     * @param line    line to check if its within the range
+     * @param column  column to see if we are within the range
+     * @return true iff the given `loc` range spans around the line/column position, inclusively.
+     */
+    private static boolean inside(ISourceLocation loc, int line, int column) {
+        if (!loc.hasLineColumn()) {
+            return false;
+        }
+
+        if (line < loc.getBeginLine() || line > loc.getEndLine()) {
+            return false;
+        }
+
+        if (line == loc.getBeginLine()) {
+            // then make sure we are not before the start
+            return loc.getBeginColumn() <= column;
+        }
+
+        if (line == loc.getEndLine()) {
+            // then make sure we are not beyond the end
+            return column <= loc.getEndColumn();
+        }
+
+        // otherwise we are in the middle of the lines 
+        // and the column is irrelevant
+
+        return true;
+    }
+    
+    /**
+     * Produces a list of trees that are "in focus" at given line and column offset (UTF-24).
+     * 
+     * This log(filesize) algorithm quickly collects the trees along a spine from the
+     * root to the smallest lexical or context-free node. The list is returned in
+     * reverse order such that you can select the "most specific" tree by starting
+     * at the start of the list.
+     * 
+     * @param tree   parse tree
+     * @param line   search term in lines (1-based)
+     * @param column search term in columns (0-based, UTF-24 codepoints)
+     * @return list of tree that are around the given line/column position, ordered from child to parent.
+     */
+    public static IList computeFocusList(ITree tree, int line, int column) {
+        var lw = IRascalValueFactory.getInstance().listWriter();
+        computeFocusList(lw, tree, line, column);
+        return lw.done();
+    }
+
+    private static boolean computeFocusList(IListWriter focus, ITree tree, int line, int column) {
+		ISourceLocation l = TreeAdapter.getLocation(tree);
+
+		if (l == null) {
+            // inside a layout, literal or character
+			return false;
+		}
+
+		if (TreeAdapter.isLexical(tree)) {
+			if (inside(l, line, column)) {
+                focus.append(tree);
+                // stop and return success
+				return true;
+			}   
+            else {
+                // stop and return failure
+                return false;
+            }
+		}
+
+		if (TreeAdapter.isAmb(tree)) {
+            // pick any tree to make the best of it.
+			return computeFocusList(focus, (ITree) tree.getAlternatives().iterator().next(), line, column);
```

Comment:
If the tree has no children (maybe it's guaranteed not to be?), an exception is thrown here